### PR TITLE
Include some references and names for strategies.

### PR DESCRIPTION
**Pull request by Nikoleta-v3**
Part of issue 225. 

Documentation for a few strategies. Including various names and author.

### CODE REVIEW

File: axelrod/strategies/punisher.py [hunk 1]
```diff
@@ -62,10 +66,11 @@ def reset(self):
 
 class InversePunisher(Player):
     """
-    A player starts by cooperating however will defect if at any point the
-    opponent has defected, but forgets after mem_length matches, with
-    1 <= mem_length <= 20 proportional to the amount of time the opponent
-    has played C. The inverse of Punisher.
+    An inverse player of Punisher.
```
Comment (drvinceknight) [inline axelrod/strategies/punisher.py]: This is not clearer than what was there. What do you mean by 'inverse'?

File: axelrod/strategies/qlearner.py [hunk 2]
```diff
@@ -7,9 +7,17 @@
 
 
 class RiskyQLearner(Player):
-    """A player who learns the best strategies through the q-learning algorithm.
+    """A player who learns the best strategies through the q-learning
+    algorithm.
+
+    Q-learning algorithm: http://www.cse.unsw.edu.au/~cs9417ml/RL1/algorithms.html
```
Comment (drvinceknight) [inline axelrod/strategies/qlearner.py]: I don't think this is necessary here.

File: axelrod/strategies/retaliate.py [hunk 3]
```diff
@@ -176,6 +196,10 @@ class LimitedRetaliate3(LimitedRetaliate):
     """
     LimitedRetaliate player with a threshold of 5 percent and a
     retaliation limit of 20.
+
+    Names:
+
+    - LimitedRetaliate3: Original strategy by Owen Campbell
```
Comment (drvinceknight) [inline axelrod/strategies/retaliate.py]: Just pinging @meatballs to check that these (see above) where indeed his.

File: axelrod/strategies/titfortat.py [hunk 4]
```diff
@@ -167,7 +200,12 @@ def strategy(opponent):
 
 
 class HardTitForTat(Player):
-    """A variant of Tit For Tat that uses a longer history for retaliation."""
+    """A variant of Tit For Tat that uses a longer history for retaliation.
+
+    Names:
+
+    - Hard Tit For Tat:
```
Comment (drvinceknight) [inline axelrod/strategies/titfortat.py]: Reference?

File: axelrod/strategies/titfortat.py [hunk 5]
```diff
@@ -194,7 +232,12 @@ def strategy(opponent):
 
 class HardTitFor2Tats(Player):
     """A variant of Tit For Two Tats that uses a longer history for
-    retaliation."""
+    retaliation.
+
+    Names:
+
+    - Hard Tit For Two Tats:
```
Comment (drvinceknight) [inline axelrod/strategies/titfortat.py]: Reference?

File: axelrod/strategies/titfortat.py [hunk 6]
```diff
@@ -297,7 +344,10 @@ class Gradual(Player):
     but after punishing enters a calming state and cooperates no matter what
     the opponent does for two rounds.
 
-    http://perso.uclouvain.be/vincent.blondel/workshops/2003/beaufils.pdf """
+    Names:
+
+    - Gradual: http://perso.uclouvain.be/vincent.blondel/workshops/2003/beaufils.pdf
```
Comment (drvinceknight) [inline axelrod/strategies/titfortat.py]: Can you reference using the bibliography.

File: docs/reference/bibliography.rst [hunk 7]
```diff
@@ -18,3 +18,8 @@ documentation.
 .. [Stewart2012] Stewart, a. J., & Plotkin, J. B. (2012). Extortion and cooperation in the Prisoner’s Dilemma. Proceedings of the National Academy of Sciences, 109(26), 10134–10135. http://doi.org/10.1073/pnas.1208087109
 .. [Szabó1992] Szabó, G., & Fáth, G. (2007). Evolutionary games on graphs. Physics Reports, 446(4-6), 97–216. http://doi.org/10.1016/j.physrep.2007.04.004
 .. [Tzafestas2000] Tzafestas, E. (2000). Toward adaptive cooperative behavior. From Animals to Animals: Proceedings of the 6th International Conference on the Simulation of Adaptive Behavior {(SAB-2000)}, 2, 334–340.
+.. [Axelrod1984]  The Evolution of Cooperation. Basic Books. ISBN 0-465-02121-2.
```
Comment (drvinceknight) [inline docs/reference/bibliography.rst]: Can you keep these in alphabetical order please.

**Comment by drvinceknight** [comment]
@Nikoleta-v3 also looks like this needs to be rebased on to current master.

File: axelrod/strategies/retaliate.py (hunk 3 above)
Comment (Nikoleta-v3) [inline axelrod/strategies/retaliate.py]: We confirmed it in PyCon. But I'll make sure to ping the author for now on.

File: axelrod/strategies/titfortat.py (hunk 5 above)
Comment (drvinceknight) [inline axelrod/strategies/titfortat.py]: Cool. In those cases can you put:

```
     - Hard Tit For Two Tats: Reference required.
```

It's still better to have the name clearly there for the case of people looking for it.

Although I think you can probably find Hard Tit For Two Tats in various papers.

File: axelrod/strategies/retaliate.py (hunk 3 above)
Comment (meatballs) [inline axelrod/strategies/retaliate.py]: Guilty

Comment (drvinceknight) [inline axelrod/strategies/retaliate.py]: > We confirmed it in PyCon. But I'll make sure to ping the author for now on.

No that's cool, hadn't realised you had chatted :) 👍

Comment (Nikoleta-v3) [inline axelrod/strategies/retaliate.py]: :smile:

File: axelrod/strategies/punisher.py [hunk 8]
```diff
@@ -62,10 +66,13 @@ def reset(self):
 
 class InversePunisher(Player):
     """
-    A player starts by cooperating however will defect if at any point the
-    opponent has defected, but forgets after mem_length matches, with
-    1 <= mem_length <= 20 proportional to the amount of time the opponent
-    has played C. The inverse of Punisher.
+    An inverse player of Punisher. Similarly the player starts by cooperating however will defect if at any point the
```
Comment (drvinceknight) [inline axelrod/strategies/punisher.py]: I think this should be: "An inverted version of Punisher."

File: axelrod/strategies/qlearner.py [hunk 9]
```diff
@@ -7,9 +7,15 @@
...
+    """A player who learns the best strategies through the q-learning
+    algorithm.
+
+    This Q learner is quick to come to conclusions and doesn't care about the
+    future.
+
+    Names:
 
-    This qlearner is quick to come to conclusions and doesn't care about the future.
+    - RiskyQLearner: Original strategy by Geraint Palmer
```
Comment (drvinceknight) [inline axelrod/strategies/qlearner.py]: Risky Q Learner

(RiskyQLearner is the name of the class)

File: axelrod/strategies/qlearner.py [hunk 10]
```diff
@@ -109,9 +115,14 @@ def reset(self):
...
 class ArrogantQLearner(RiskyQLearner):
-    """A player who learns the best strategies through the q-learning algorithm.
+    """A player who learns the best strategies through the q-learning
+    algorithm.
 
     This Q learner jumps to quick conclusions and care about the future.
+
+    Names:
+
+    - ArrogantQLearner: Original strategy by Geraint Palmer
```
Comment (drvinceknight) [inline axelrod/strategies/qlearner.py]: Arrogant Q Learner

File: axelrod/strategies/qlearner.py [hunk 11]
```diff
@@ -123,6 +134,10 @@ class HesitantQLearner(RiskyQLearner):
     """A player who learns the best strategies through the q-learning algorithm.
 
     This Q learner is slower to come to conclusions and does not look ahead much.
+
+    Names:
+
+    - HesitantQLearner: Original strategy by Geraint Palmer
```
Comment (drvinceknight) [inline axelrod/strategies/qlearner.py]: Hesitant Q Learner

File: axelrod/strategies/qlearner.py [hunk 12]
```diff
@@ -133,7 +148,12 @@ class HesitantQLearner(RiskyQLearner):
 class CautiousQLearner(RiskyQLearner):
     """A player who learns the best strategies through the q-learning algorithm.
 
-    This Q learner is slower to come to conclusions and wants to look ahead more.
+    This Q learner is slower to come to conclusions and wants to look ahead
+    more.
+
+    Names:
+
+    - CautiousQLearner: Original strategy by Geraint Palmer
```
Comment (drvinceknight) [inline axelrod/strategies/qlearner.py]: Cautious Q Learner

File: axelrod/strategies/sequence_player.py [hunk 13]
```diff
@@ -37,9 +38,14 @@ def reset(self):
...
     A player who cooperates or defects according to the Thue-Morse sequence.
-
     The first few terms of the Thue-Morse sequence are:
     0 1 1 0 1 0 0 1 1 0 0 1 0 1 1 0 . . .
+
+    Thue-Morse sequence: http://mathworld.wolfram.com/Thue-MorseSequence.html
+
+    Names:
+
+    - ThueMorse: Original by Geraint Palmer
```
Comment (drvinceknight) [inline axelrod/strategies/sequence_player.py]: Thue Morse

File: axelrod/strategies/sequence_player.py [hunk 14]
```diff
@@ -59,8 +65,12 @@ def __init__(self):
 
 
 class ThueMorseInverse(ThueMorse):
-    """A player who defects or cooperates according to the Thue-Morse sequence
-    (Inverse of ThueMorse)."""
+    """ A player who plays the inverse of the Thue-Morse sequence.
+
+    Names:
+
+    - ThueMorse: Original by Geraint Palmer
```
Comment (drvinceknight) [inline axelrod/strategies/sequence_player.py]: Inverse Thue Morse

File: axelrod/strategies/titfortat.py [hunk 15]
```diff
@@ -62,7 +68,13 @@ def strategy(opponent):
 
 
 class TwoTitsForTat(Player):
-    """A player starts by cooperating and replies to each defect by two defections."""
+    """A player starts by cooperating and replies to each defect by two
+    defections.
+
+    Names:
+
+    - Tit for two Tats: [Axelrod1984]_
```
Comment (drvinceknight) [inline axelrod/strategies/titfortat.py]: "Two tits for Tat" I assume?

File: axelrod/strategies/titfortat.py [hunk 16]
```diff
@@ -148,7 +176,12 @@ def strategy(opponent):
 
 class AntiTitForTat(Player):
     """A strategy that plays the opposite of the opponents previous move.
-    This is similar to BULLY above, except that the first move is cooperation."""
+    This is similar to BULLY above, except that the first move is cooperation.
```
Comment (drvinceknight) [inline axelrod/strategies/titfortat.py]: Can you get rid of "above" please, so:

```
This is similar to Bully, except that...
```

File: axelrod/strategies/titfortat.py (hunk 5 above)
Comment (Nikoleta-v3) [inline axelrod/strategies/titfortat.py]: You are correct. I am leaving 'Reference required' for now though.

File: docs/reference/bibliography.rst [hunk 17]
```diff
@@ -7,14 +7,20 @@ This is a collection of various bibliographic items referenced in the
 documentation.
 
 .. [Axelrod1980] Axelrod, R. (1980). Effective Choice in the Prisoner’s Dilemma. Journal of Conflict Resolution, 24(1), 3–25.
+.. [Axelrod1984]  The Evolution of Cooperation. Basic Books. ISBN 0-465-02121-2.
+.. [Axelrod1995] Wu, J. and Axelrod, R. (1995) ‘How to cope with noise in the Iterated prisoner’s dilemma’, Journal of Conflict Resolution, 39(1), pp. 183–189. doi: 10.1177/0022002795039001008.
```
Comment (drvinceknight) [inline docs/reference/bibliography.rst]: No need for the `'`

File: docs/reference/bibliography.rst [hunk 18]
```diff
@@ -7,14 +7,20 @@ This is a collection of various bibliographic items referenced in the
...
 .. [Banks1980] Banks, J. S., & Sundaram, R. K. (1990). Repeated games, finite automata, and complexity. Games and Economic Behavior, 2(2), 97–117.  http://doi.org/10.1016/0899-8256(90)90024-O
+.. [Beaufils1997] Beaufils, B. and Delahaye, J. (1997) ‘Our Meeting With Gradual: A Good Strategy For The Iterated Prisoner’s Dilemma’. http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.42.4041
```
Comment (drvinceknight) [inline docs/reference/bibliography.rst]: Same comment.

File: docs/reference/bibliography.rst [hunk 19]
```diff
@@ -7,14 +7,20 @@ This is a collection of various bibliographic items referenced in the
...
 .. [Berg2015] Berg, P. Van Den, & Weissing, F. J. (2015). The importance of mechanisms for the evolution of cooperation. Proceedings of the Royal Society B-Biological Sciences, 282.
+.. [Hilde2013] Hilbe, C., Nowak, M.A. and Traulsen, A. (2013) ‘Adaptive dynamics of extortion and compliance’, PLoS ONE, 8(11), p. e77886. doi: 10.1371/journal.pone.0077886.
```
Comment (drvinceknight) [inline docs/reference/bibliography.rst]: Same comment.

**Comment by Nikoleta-v3** [comment]
@drvinceknight I believe I addressed all your comments. I also did rebase. Waiting for further comments from @meatballs :smile:

File: axelrod/strategies/sequence_player.py (hunk 14 above)
Comment (drvinceknight) [inline axelrod/strategies/sequence_player.py]: You missed this one @Nikoleta-v3. Needs to be "Inverse Thue Morse". (Currently both the Thue Morse strategies have the same name).

**Comment by drvinceknight** [comment]
This is fine by me. I've edited the name of the PR so that it's verbose.

File: axelrod/strategies/punisher.py [hunk 20]
```diff
@@ -62,10 +66,13 @@ def reset(self):
 
 class InversePunisher(Player):
     """
-    A player starts by cooperating however will defect if at any point the
-    opponent has defected, but forgets after mem_length matches, with
-    1 <= mem_length <= 20 proportional to the amount of time the opponent
-    has played C. The inverse of Punisher.
+    An inverted version of Punisher. Similarly the player starts by cooperating however will defect if at any point the
+    opponent has defected, but forgets after mem_length matches, with 1 <= mem_length <= 20. This time mem_lenght is
```
Comment (meatballs) [inline axelrod/strategies/punisher.py]: typo: 'mem_lenght' should be 'mem_length'

File: axelrod/strategies/qlearner.py [hunk 21]
```diff
@@ -109,9 +115,14 @@ def reset(self):
 
 
 class ArrogantQLearner(RiskyQLearner):
-    """A player who learns the best strategies through the q-learning algorithm.
+    """A player who learns the best strategies through the q-learning
+    algorithm.
 
     This Q learner jumps to quick conclusions and care about the future.
```
Comment (meatballs) [inline axelrod/strategies/qlearner.py]: typo: 'care' should be 'cares'

File: axelrod/strategies/titfortat.py [hunk 22]
```diff
@@ -84,7 +96,13 @@ class Bully(Player):
     """A player that behaves opposite to Tit For Tat, including first move.
 
     Starts by defecting and then does the opposite of opponent's previous move.
-    This is the complete opposite of TIT FOR TAT, also called BULLY in literature.
+    This is the complete opposite of Tit For Tat, also called BULLY in
+    literature.
```
Comment (meatballs) [inline axelrod/strategies/titfortat.py]: typo: should be 'in THE literature'

File: axelrod/strategies/titfortat.py [hunk 23]
```diff
@@ -128,7 +151,12 @@ def strategy(self, opponent):
 
 
 class SuspiciousTitForTat(Player):
-    """A TFT that initially defects."""
+    """A variant of Tit For Tats that starts off with a defection.
```
Comment (meatballs) [inline axelrod/strategies/titfortat.py]: typo: 'Tats' should be 'Tat'

Comment (Nikoleta-v3) [inline axelrod/strategies/titfortat.py]: All done @meatballs :smile:

**Pull request merged**